### PR TITLE
DOC-5995: Added Ledger Recovery Configuration fields

### PR DESCRIPTION
**Pull request by AnnaK-R3**
Preview: https://anna.preview.docs.r3.com/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.html#enterpriseconfiguration

### CODE REVIEW

**Review by balwantkothari — COMMENTED** [review]
In "LedgerRecoveryConfiguration" L is small case it should be "ledgerRecoveryConfiguration"

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 1]
```diff
@@ -387,6 +387,110 @@ Allows fine-grained controls of various features only available in the enterpris
 
       - The size of the in-memory Artemis buffer for messages, in bytes. Note that there is a lower bound to the buffer size, which is calculated based on the maximum message size of the network parameters to ensure messages of any allowed size can be stored successfully. As a result, any value lower than this bound will be ignored with the appropriate logging. This bound is also used as the default, if no value is specified.
 
+* `ledgerRecoveryConfiguration`
+
+    * This configuration allows you to tailor ledger recovery behavior for your Corda node. It offers flexibility in defining
+    parameters related to key pair pre-generation, backup intervals, and confidential identity backup options.
```
Comment (adelel1) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:393]: Just thinking, as this is a fairly big topic. What about adding something like the following to the end of this description. Where you link to the doc that Jose and Balwant wrote.
Please see XXXX for a detailed description of ledger recovery that makes use of this configuration.

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 2]
```diff
@@ -387,6 +387,110 @@ Allows fine-grained controls of various features only available in the enterpris
...
+        * If the parameter is specified in more than one place, then the recovery flow parameter overrides the configuration
+        parameter which overrides the network parameter. If the node is on a network not using CENM 1.6 and the parameter
+        is not specified in either configuration or directly on the ledger recovery flow, then an exception is raised
+        when starting the recovery flow.
+        * There is no default for this parameter. If it is not specified in any of the above
+        ways, an exception is raised.
+        * The duration is specified in the Human-Optimized Config Object Notation (HOCON) format with suffixes of `h` (hours),
+        `m` (minutes) and `s` (seconds), for example, `1h` means one hour. For additional information on the HOCON duration
+        format parsing, see the [HOCON duration format file](https://github.com/lightbend/config/blob/master/HOCON.md) on GitHub.
+        * *Default:* null
```
Comment (adelel1) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:439]: There is no default. If the recovery flow is run then recoveryMaximumBackupInterval must be set by one of the ways described above.

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 3]
```diff
@@ -387,6 +387,110 @@ Allows fine-grained controls of various features only available in the enterpris
...
+
+        * If the parameter is specified in more than one place, then the configuration parameter overrides the network parameter.
+        * If the node is on a network not using CENM 1.6 and the parameter is not specified in configuration, then an exception
+        is raised if either `noOfPreGeneratedKeys` or `noOfPreGeneratedKeysWithCerts` is greater than 0. In other words, if
+        you want to regenerate keys, then you must have the `confidentialIdentityMinimumBackupInterval` defined. There is
+        no default for this parameter.
+        * The duration is specified in the Human-Optimized Config Object Notation (HOCON) format with suffixes of `h` (hours),
+        `m` (minutes) and `s` (seconds), for example, `1h` means one hour. For additional information on the HOCON duration
+        format parsing, see the [HOCON duration format file](https://github.com/lightbend/config/blob/master/HOCON.md) on GitHub.
+        * *Default:* null
```
Comment (adelel1) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:457]: If either noOfPreGeneratedKeys or noOfPreGeneratedKeysWithCerts is greater than zero then confidentialIdentityMinimumBackupInterval must be set by one of the ways described above.

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 4]
```diff
@@ -387,6 +387,110 @@ Allows fine-grained controls of various features only available in the enterpris
...
+
+    - `canProvideNonBackedUpKeyPair`
+
+        * Determines whether the node can provide non-backed-up key pairs if backed-up keys pairs are exhausted, or if CIs
+        are used but both `noOfPreGeneratedKeys` and `noOfPreGeneratedKeysWithCerts` are zero.
```
Comment (adelel1) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:462]: shouldn't it be....
are used and both

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 5]
```diff
@@ -387,6 +387,110 @@ Allows fine-grained controls of various features only available in the enterpris
...
+
+    - `bufferSizeForKeys`
+
+        * By configuring this property, you can control the number of pre-generated keys that are readily available in memory.
+        * *Default:* null
+
+    - `bufferSizeForKeys` with certificates
+
+        * By configuring this property, you can control the number of pre-generated keys with certificates that are readily available in memory.
+        * *Default:* null
```
Comment (adelel1) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:473]: Default: 0

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 6]
```diff
@@ -387,6 +387,110 @@ Allows fine-grained controls of various features only available in the enterpris
...
+    - `canProvideNonBackedUpKeyPair`
+
+        * Determines whether the node can provide non-backed-up key pairs if backed-up keys pairs are exhausted, or if CIs
+        are used but both `noOfPreGeneratedKeys` and `noOfPreGeneratedKeysWithCerts` are zero.
+        * *Default:* true
+
+    - `bufferSizeForKeys`
+
+        * By configuring this property, you can control the number of pre-generated keys that are readily available in memory.
+        * *Default:* null
```
Comment (adelel1) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:468]: Default: 0

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 2 above)
Comment (AnnaK-R3) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:439]: Removed the default.

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 3 above)
Comment (AnnaK-R3) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:457]: Updated

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 4 above)
Comment (AnnaK-R3) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:462]: updated

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 1 above)
Comment (AnnaK-R3) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:393]: Ok, I'll add the reference later and leave this PR open until I merge the ledger recovery one as I still haven't decided on the location of this content.

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 3 above)
Comment (adelel1) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md]: I meant to replace Default: null with:

Default: If either noOfPreGeneratedKeys or noOfPreGeneratedKeysWithCerts is greater than zero then confidentialIdentityMinimumBackupInterval must be set by one of the ways described above.

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 2 above)
Comment (adelel1) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md]: I mean to replace Default: null with...
Default: If the recovery flow is run then recoveryMaximumBackupInterval must be set by one of the ways described above.

**Review by adelel1 — DISMISSED** [review]

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 2 above)
Comment (AnnaK-R3) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md]: Replaced

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md (hunk 1 above)
Comment (AnnaK-R3) [reply content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:394]: Added link

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 7]
```diff
@@ -388,6 +388,104 @@ Allows fine-grained controls of various features only available in the enterpris
...
 
+* `ledgerRecoveryConfiguration`
+
+    * This configuration allows you to tailor Ledger Recovery behavior for your Corda node. It offers flexibility in defining parameters related to key pair pre-generation, backup intervals, and confidential identity backup options. For a detailed description of Ledger Recovery that uses this configuration, see [Ledger Recovery]({{< relref "../collaborative-recovery/ledger-recovery/overview.md" >}}).
+
+    - `noOfPreGeneratedKeys`
+
+        * This property specifies the number of pre-generated keys used for confidential identities, indicating the count
+        of keys that will be backed up in the database. It represents the pre-generated count of so-called new
+        confidential identities, that is, those that don’t have a certificate.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:399]:
```suggestion
        confidential identities; that is, those that don’t have a certificate.
```

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 8]
```diff
@@ -388,6 +388,104 @@ Allows fine-grained controls of various features only available in the enterpris
...
+        * The default for this property is 0 which means that if you have not changes it and if you are using new confidential identities and  (and have not changed `canProvideNonBackedUpKeyPair` default), then you receive a confidential identity created on the fly, not a pre-generated one. If you have changed `canProvideNonBackedUpKeyPair` to `false` and if there are no backed up keys to return, then an exception is raised.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:400]:
```suggestion
        * The default for this property is 0, which means that:
           * if you have not changed it
           * if you are using new confidential identities, and
           * if you have not changed `canProvideNonBackedUpKeyPair` from its default value
           then you receive a confidential identity created on the fly, not a pre-generated one. If you have changed `canProvideNonBackedUpKeyPair` to `false` and if there are no backed-up keys to return, then an exception is raised.
```


Hope this format works....

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 9]
```diff
@@ -388,6 +388,104 @@ Allows fine-grained controls of various features only available in the enterpris
...
+        * *Default:* 0
+
+    - `noOfPreGeneratedKeysWithCerts`
+
+        * This property specifies the number of pre-generated keys with certificates used for confidential identities,
+        indicating the count of keys that will be backed up in the database. It represents the pre-generated count
+        of so-called old confidential identities, that is, those that have a certificate.
+        * The default for this property is 0 which means if you have not changed it and if you are using new confidential identities (and have not changed `canProvideNonBackedUpKeyPair` default), then you receive a confidential identity created on the fly, not a pre-generated one. If you have changed `canProvideNonBackedUpKeyPair` to `false` and if there are no backed up keys to return, then an exception is raised.
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:408]:
```suggestion
        * The default for this property is 0, which means if:
          * you have not changed it
          * you are using new confidential identities, and 
          * you have not changed `canProvideNonBackedUpKeyPair` from its default value)
          then you receive a confidential identity created on the fly, not a pre-generated one. If you have changed `canProvideNonBackedUpKeyPair` to `false` and if there are no backed-up keys to return, then an exception is raised.
```

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 10]
```diff
@@ -388,6 +388,104 @@ Allows fine-grained controls of various features only available in the enterpris
...
+          * Directly as a parameter on the recovery flow
+
+        * If the parameter is specified in more than one place, then the recovery flow parameter overrides the configuration
+        parameter which overrides the network parameter. If the node is on a network not using CENM 1.6 and the parameter
+        is not specified in either configuration or directly on the ledger recovery flow, then an exception is raised
+        when starting the recovery flow.
+        * There is no default for this parameter. If it is not specified in any of the above
+        ways, an exception is raised.
+        * The duration is specified in the Human-Optimized Config Object Notation (HOCON) format with suffixes of `h` (hours),
+        `m` (minutes) and `s` (seconds), for example, `1h` means one hour. For additional information on the HOCON duration
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:431]:
```suggestion
        `m` (minutes) and `s` (seconds); for example, `1h` means one hour. For additional information on the HOCON duration
```

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 11]
```diff
@@ -388,6 +388,104 @@ Allows fine-grained controls of various features only available in the enterpris
...
+          * Network parameters (if node is on a network using CENM 1.6 or later)
+          * Configuration parameter (this parameter)
+
+        * If the parameter is specified in more than one place, then the configuration parameter overrides the network parameter.
+        * If the node is on a network not using CENM 1.6 and the parameter is not specified in configuration, then an exception
+        is raised if either `noOfPreGeneratedKeys` or `noOfPreGeneratedKeysWithCerts` is greater than 0. In other words, if
+        you want to regenerate keys, then you must have the `confidentialIdentityMinimumBackupInterval` defined. There is
+        no default for this parameter.
+        * The duration is specified in the Human-Optimized Config Object Notation (HOCON) format with suffixes of `h` (hours),
+        `m` (minutes) and `s` (seconds), for example, `1h` means one hour. For additional information on the HOCON duration
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:449]:
```suggestion
        `m` (minutes) and `s` (seconds); for example, `1h` means one hour. For additional information on the HOCON duration
```

File: content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md [hunk 12]
```diff
@@ -388,6 +388,104 @@ Allows fine-grained controls of various features only available in the enterpris
...
+        * By configuring this property, you can control the number of pre-generated keys that are readily available in memory.
+        * *Default:* 0
+
+    - `bufferSizeForKeys` with certificates
+
+        * By configuring this property, you can control the number of pre-generated keys with certificates that are readily available in memory.
+        * *Default:* 0
+
+    * To configure the `ledgerRecoveryConfiguration` for your Corda Enterprise node, modify the properties according to the
+    requirements in your node's configuration file, for example:
```
Comment (paulmoloneyr3) [inline content/en/platform/corda/4.11/enterprise/node/setup/corda-configuration-fields.md:471]:
```suggestion
    requirements in your node's configuration file; for example:
```

**Review by paulmoloneyr3 — CHANGES_REQUESTED** [review]

**Review by paulmoloneyr3 — APPROVED** [review]

**Pull request merged**